source/declarations.tex

```diff
@@ -4239,3 +4240,414 @@
+\item For a contract condition of a public member function,
+no member of \tcode{C} or of an enclosing class of \tcode{C} is accessible
+unless it is a public member of \tcode{C},
+or a member of a base class
+accessible as a public member of \tcode{C}\iref{class.access.base}.
+\item For a contract condition of a protected member function,
+no member of \tcode{C} or of an enclosing class of \tcode{C} is accessible
+unless it is a public or protected member of \tcode{C},
+or a member of a base class
+accessible as a public or protected member of \tcode{C}.
+\end{itemize}
+For names appearing in a contract condition of a non-member function,
+friendship is not considered.
+\begin{example}
+\begin{codeblock}
+class X {
+public:
+  int v() const;
+  void f() [[expects: x > 0]];                  // error: \tcode{x} is private
+  void g() [[expects: v() > 0]];                // OK
+  friend void r(int z) [[expects: z > 0]];      // OK
+  friend void s(int z) [[expects: z > x]];      // error: \tcode{x} is private
+protected:
+  int w();
+  void h() [[expects: x > 0]];                  // error: \tcode{x} is private
+  void i() [[ensures: y > 0]];                  // OK
+  void j() [[ensures: w() > 0]];                // OK
+  int y;
+private:
+  void k() [[expects: x > 0]];                  // OK
+  int x;
+};
+
+class Y : public X {
+public:
+  void a() [[expects: v() > 0]];                // OK
+  void b() [[ensures: w() > 0]];                // error: \tcode{w} is protected
+protected:
+  void c() [[expects: w() > 0]];                // OK
+};
+\end{codeblock}
+\end{example}
+
+\pnum
+A precondition is checked by evaluating its predicate
+immediately before starting evaluation of the function body.
+\begin{note}
+The function body includes
+the \grammarterm{function-try-block}\iref{except} and
+the \grammarterm{ctor-initializer}\iref{class.base.init}.
+\end{note}
+A postcondition is checked by evaluating its predicate
+immediately before returning control to the caller of the function.
+\begin{note}
+The lifetime of local variables and temporaries has ended.
+Exiting via an exception or via \tcode{longjmp}\iref{csetjmp.syn}
+is not considered returning control to the caller of the function.
+\end{note}
+
+\pnum
+If a function has multiple preconditions,
+their evaluation (if any) will be performed
+in the order they appear lexically.
+If a function has multiple postconditions,
+their evaluation (if any) will be performed
+in the order they appear lexically.
+\begin{example}
+\begin{codeblock}
+void f(int * p)
+  [[expects: p != nullptr]]                     // \#1
+  [[ensures: *p == 1]]                          // \#3
+  [[expects: *p == 0]]                          // \#2
+{
+  *p = 1;
+}
+\end{codeblock}
+\end{example}
+
+\pnum
+An assertion is checked by evaluating its predicate
+as part of the evaluation of the null statement\iref{stmt.expr}
+it applies to.
+
+\pnum
+The only side effects of a predicate
```

Comment:
This is more about predicates in the abstract; maybe this should go after p5, where we talk about contextual conversion to bool (and now, about being potentially-evaluated)?

Reply:
I disagree.
The previous paragraphs talk about syntactic restrictions, e.g. the funny access rules and where the attributes may be applied. When then define "checked" in terms of "evaluated". And then we discuss constraints on the specific evaluation, such as the paragraph in question and the next one, which talks about prohibited things in the evaluation of the function body (outside of contracts).

Reply:
OK